sdks/ios/ZelloChannelKit/ZelloChannelKit/network/ZCCSocket.m

```diff
@@ -675,6 +753,71 @@ - (void)handleTextMessage:(NSDictionary *)encoded original:(NSString *)original
   }];
 }
 
+- (void)handleImage:(NSDictionary *)encoded original:(NSString *)original {
+  ZCCImageHeader *header = [[ZCCImageHeader alloc] init];
+  id channel = encoded[ZCCChannelNameKey];
+  if ([channel isKindOfClass:[NSString class]]) {
+    header.channel = channel;
+  }
+  id sender = encoded[ZCCFromUserKey];
+  if ([sender isKindOfClass:[NSString class]]) {
+    header.sender = sender;
+  }
+  id imageId = encoded[ZCCMessageIDKey];
+  if ([imageId isKindOfClass:[NSNumber class]]) {
+    long long imageIdValue = [imageId longLongValue];
+    if (imageIdValue < 0 || imageIdValue > UINT32_MAX) {
+      [self reportInvalidStringMessage:original];
+      return;
+    }
+    header.imageId = (NSUInteger)imageIdValue;
+  }
+  id source = encoded[ZCCImageSourceKey];
+  if ([source isKindOfClass:[NSString class]]) {
+    header.source = source;
+  }
+  id type = encoded[ZCCStreamTypeKey];
+  if ([type isKindOfClass:[NSString class]]) {
+    if ([type isEqualToString:@"jpeg"]) {
+      header.imageType = ZCCImageTypeJPEG;
+    } else {
+      header.imageType = ZCCImageTypeUnkown;
+    }
+  } else {
+    [self reportInvalidStringMessage:original];
+    return;
+  }
+  id height = encoded[ZCCImageHeightKey];
+  if ([height isKindOfClass:[NSNumber class]]) {
+    long long heightValue = [height longLongValue];
+    if (heightValue < 0 && heightValue > INT32_MAX) {
+      [self reportInvalidStringMessage:original];
+      return;
+    }
+    header.height = (NSInteger)heightValue;
+  } else {
+    [self reportInvalidStringMessage:original];
+    return;
+  }
+  id width = encoded[ZCCImageWidthKey];
+  if ([width isKindOfClass:[NSNumber class]]) {
+    long long widthValue = [width longLongValue];
+    if (widthValue < 0 || widthValue > INT32_MAX) {
+      [self reportInvalidStringMessage:original];
+      return;
+    }
+    header.width = widthValue;
+  } else {
+    [self reportInvalidStringMessage:original];
+    return;
+  }
+
+  id<ZCCSocketDelegate> delegate = self.delegate;
+  [self.delegateRunner runAsync:^{
+    [delegate socket:self didReceiveImageHeader:header];
+  }];
+}
+
 - (void)reportInvalidStringMessage:(NSString *)message {
```

Comment:
this is not very informative so the above calls to it don't really inform us.

We need three things to diagnose problems in the wild with JSON payloads.

- [x] The payload. (Good job assuming that is "message")
- [ ] The troubled keypath
- [ ] A classification for the error.

In our logic checks above we have all this information... so it should not be a giant stretch to inform future us as to what issues are without parsing JSON bodies as a whole.

Reply:
That's a really good point. I think it's outside the scope of this feature, but I'll implement better error reporting in its own pull request.